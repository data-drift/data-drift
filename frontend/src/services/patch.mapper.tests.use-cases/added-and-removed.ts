import { Row } from "../../components/Table/Table";

export const patch =
  "@@ -2,9 +2,9 @@ unique_key,name,date,age\n 2022-12-Alice,Alice,2022-12,25\n 2023-01-Bob,Bob,2023-01,30\n 2023-01-Charlie,Charlie,2023-01,36\n+2023-01-Charline,Charline,2023-01,42\n 2023-02-Antoine,Antoine,2023-02,40\n-2023-02-Didier,Didier,2023-02,40\n+2023-02-Didier,Didier,2023-02,43\n 2023-02-Philipe,Philipe,2023-02,42\n-2023-03-Clement,Clement,2023-03,45\n 2023-03-Cyril,Cyril,2023-03,45\n 2023-03-Victor,Victor,2023-03,46";

export const expectedOldData: Row[] = [
  {
    isEmphasized: false,
    data: [
      { value: "2022-12-Alice", type: "string" },
      { value: "Alice", type: "string" },
      { value: "2022-12", type: "string" },
      { value: "25", type: "number" },
    ],
  },
  {
    isEmphasized: false,
    data: [
      { value: "2023-01-Bob", type: "string" },
      { value: "Bob", type: "string" },
      { value: "2023-01", type: "string" },
      { value: "30", type: "number" },
    ],
  },
  {
    isEmphasized: false,
    data: [
      { value: "2023-01-Charlie", type: "string" },
      { value: "Charlie", type: "string" },
      { value: "2023-01", type: "string" },
      { value: "36", type: "number" },
    ],
  },
  {
    isEmphasized: false,
    data: [{ value: "_" }, { value: "_" }, { value: "_" }, { value: "_" }],
  }, // empty line here
  {
    isEmphasized: false,
    data: [
      { value: "2023-02-Antoine", type: "string" },
      { value: "Antoine", type: "string" },
      { value: "2023-02", type: "string" },
      { value: "40", type: "number" },
    ],
  },
  {
    isEmphasized: true,
    data: [
      { value: "2023-02-Didier", type: "string" },
      { value: "Didier", type: "string" },
      { value: "2023-02", type: "string" },
      { value: "40", type: "number", isEmphasized: true },
    ],
  },
  {
    isEmphasized: false,
    data: [
      { value: "2023-02-Philipe", type: "string" },
      { value: "Philipe", type: "string" },
      { value: "2023-02", type: "string" },
      { value: "42", type: "number" },
    ],
  },
  {
    isEmphasized: true,
    data: [
      { value: "2023-03-Clement", type: "string" },
      { value: "Clement", type: "string" },
      { value: "2023-03", type: "string" },
      { value: "45", type: "number" },
    ],
  },
  {
    isEmphasized: false,
    data: [
      { value: "2023-03-Cyril", type: "string" },
      { value: "Cyril", type: "string" },
      { value: "2023-03", type: "string" },
      { value: "45", type: "number" },
    ],
  },
  {
    isEmphasized: false,
    data: [
      { value: "2023-03-Victor", type: "string" },
      { value: "Victor", type: "string" },
      { value: "2023-03", type: "string" },
      { value: "46", type: "number" },
    ],
  },
];

export const expectedNewData: Row[] = [
  {
    isEmphasized: false,
    data: [
      { value: "2022-12-Alice", type: "string" },
      { value: "Alice", type: "string" },
      { value: "2022-12", type: "string" },
      { value: "25", type: "number" },
    ],
  },
  {
    isEmphasized: false,
    data: [
      { value: "2023-01-Bob", type: "string" },
      { value: "Bob", type: "string" },
      { value: "2023-01", type: "string" },
      { value: "30", type: "number" },
    ],
  },
  {
    isEmphasized: false,
    data: [
      { value: "2023-01-Charlie", type: "string" },
      { value: "Charlie", type: "string" },
      { value: "2023-01", type: "string" },
      { value: "36", type: "number" },
    ],
  },
  {
    isEmphasized: true,
    data: [
      { value: "2023-01-Charline", type: "string" },
      { value: "Charline", type: "string" },
      { value: "2023-01", type: "string" },
      { value: "42", type: "number" },
    ],
  },
  {
    isEmphasized: false,
    data: [
      { value: "2023-02-Antoine", type: "string" },
      { value: "Antoine", type: "string" },
      { value: "2023-02", type: "string" },
      { value: "40", type: "number" },
    ],
  },
  {
    isEmphasized: true,
    data: [
      { value: "2023-02-Didier", type: "string" },
      { value: "Didier", type: "string" },
      { value: "2023-02", type: "string" },
      { value: "43", type: "number", isEmphasized: true },
    ],
  },
  {
    isEmphasized: false,
    data: [
      { value: "2023-02-Philipe", type: "string" },
      { value: "Philipe", type: "string" },
      { value: "2023-02", type: "string" },
      { value: "42", type: "number" },
    ],
  },
  {
    isEmphasized: false,
    data: [{ value: "_" }, { value: "_" }, { value: "_" }, { value: "_" }],
  }, // empty line here
  {
    isEmphasized: false,
    data: [
      { value: "2023-03-Cyril", type: "string" },
      { value: "Cyril", type: "string" },
      { value: "2023-03", type: "string" },
      { value: "45", type: "number" },
    ],
  },
  {
    isEmphasized: false,
    data: [
      { value: "2023-03-Victor", type: "string" },
      { value: "Victor", type: "string" },
      { value: "2023-03", type: "string" },
      { value: "46", type: "number" },
    ],
  },
];
