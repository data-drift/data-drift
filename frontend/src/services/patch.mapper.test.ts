import { parsePatch } from "./patch.mapper";
import * as addedAndRemoved from "./patch.mapper.tests.use-cases/added-and-removed.ts";

describe("parsePatch", () => {
  it("should detect added and removed lines with isEmphasized on the rows", () => {
    const patch = addedAndRemoved.patch;
    const { oldData, newData } = parsePatch(patch);

    expect(oldData.diffType).toBe("removed");
    expect(oldData.headers).toEqual(["unique_key", "name", "date", "age"]);

    expect(oldData.data).toEqual(addedAndRemoved.expectedOldData);

    expect(newData.diffType).toBe("added");
    expect(newData.headers).toEqual(["unique_key", "name", "date", "age"]);
    expect(newData.data).toEqual(addedAndRemoved.expectedNewData);
  });

  it("should parse the headers in the patch header", () => {
    const patch =
      "@@ -2,9 +2,9 @@ unique_key,name,date,age\n 2022-12-Alice,Alice,2022-12,25";
    const { oldData, newData } = parsePatch(patch);

    expect(oldData.headers).toEqual(["unique_key", "name", "date", "age"]);
    expect(newData.headers).toEqual(["unique_key", "name", "date", "age"]);
  });

  it("should parse the headers in the first line", () => {
    const patch =
      "@@ -2,9 +2,9 @@\n unique_key,name,date,age\n 2022-12-Alice,Alice,2022-12,25";
    const { oldData, newData } = parsePatch(patch);

    expect(oldData.headers).toEqual(["unique_key", "name", "date", "age"]);
    expect(newData.headers).toEqual(["unique_key", "name", "date", "age"]);
  });

  it("should detect modified lines with isEmphasized on the rows", () => {
    const patch =
      "@@ -2,9 +2,9 @@ unique_key,name,date,age\n-2022-12-Alice,Alice,2022-12,25\n+2022-12-Alice,Alice,2022-12,26";
    const { oldData, newData } = parsePatch(patch);
    expect(oldData.data[0].data[3]).toStrictEqual({
      value: "25",
      isEmphasized: true,
    });
    expect(newData.data[0].data[3]).toStrictEqual({
      value: "26",
      isEmphasized: true,
    });
  });
});
