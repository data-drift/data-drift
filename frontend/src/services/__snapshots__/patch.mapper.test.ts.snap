// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parsePatch should handle large datasets 1`] = `
{
  "newData": {
    "data": [
      {
        "data": [
          {
            "value": "1d6b2882-9a26-41d6-a976-dff75dfdd5e2",
          },
          {
            "value": "2004-05-11",
          },
          {
            "value": "3.93",
          },
          {
            "value": "PT",
          },
          {
            "value": "Category C",
          },
        ],
        "isEmphasized": false,
      },
      {
        "data": [
          {
            "value": "1d6b2f8c-13c7-4c33-9835-d7e9e5448f81",
          },
          {
            "value": "2012-12-16",
          },
          {
            "value": "2.18",
          },
          {
            "value": "AF",
          },
          {
            "value": "Category C",
          },
        ],
        "isEmphasized": false,
      },
      {
        "data": [
          {
            "value": "1d6b412b-e1be-4566-b27f-d6d2591be73e",
          },
          {
            "value": "2019-11-01",
          },
          {
            "value": "4.0",
          },
          {
            "value": "SI",
          },
          {
            "value": "Category A",
          },
        ],
        "isEmphasized": false,
      },
      {
        "data": [
          {
            "value": "1d6c2c23-39cb-418a-9e09-e22b8ca3601f",
          },
          {
            "value": "1993-10-25",
          },
          {
            "isEmphasized": true,
            "value": "7.22",
          },
          {
            "value": "PE",
          },
          {
            "value": "Category B",
          },
        ],
        "isEmphasized": true,
      },
      {
        "data": [
          {
            "value": "1d6c6e9c-3f2b-4f36-af8f-4ece80e1c456",
          },
          {
            "value": "2021-11-07",
          },
          {
            "value": "1.72",
          },
          {
            "value": "ME",
          },
          {
            "value": "Category C",
          },
        ],
        "isEmphasized": false,
      },
      {
        "data": [
          {
            "value": "1d6cd9a0-4f30-452f-8dda-34777e2814e0",
          },
          {
            "value": "2002-01-04",
          },
          {
            "value": "7.89",
          },
          {
            "value": "AM",
          },
          {
            "value": "Category C",
          },
        ],
        "isEmphasized": false,
      },
      {
        "data": [
          {
            "value": "1d6e65d0-0b9b-4b76-9517-ed65dd038b14",
          },
          {
            "value": "2009-01-25",
          },
          {
            "value": "4.63",
          },
          {
            "value": "GW",
          },
          {
            "value": "Category C",
          },
        ],
        "isEmphasized": false,
      },
      {
        "data": [],
        "isEllipsis": true,
      },
      {
        "data": [
          {
            "value": "3d1634ea-2648-41cf-8bc4-02063d57d6d4",
          },
          {
            "value": "2015-05-19",
          },
          {
            "value": "8.43",
          },
          {
            "value": "KE",
          },
          {
            "value": "Category A",
          },
        ],
        "isEmphasized": false,
      },
      {
        "data": [
          {
            "value": "3d16de2d-c626-46b7-90bb-211be6e38f31",
          },
          {
            "value": "2004-02-27",
          },
          {
            "value": "0.69",
          },
          {
            "value": "JO",
          },
          {
            "value": "Category A",
          },
        ],
        "isEmphasized": false,
      },
      {
        "data": [
          {
            "value": "3d16e1f2-a2fc-4b6a-a00b-fbc4e16aa485",
          },
          {
            "value": "2012-05-20",
          },
          {
            "value": "2.17",
          },
          {
            "value": "CZ",
          },
          {
            "value": "Category C",
          },
        ],
        "isEmphasized": false,
      },
    ],
    "diffType": "added",
    "headers": [
      "unique_key",
      "date",
      "metric_value",
      "country_code",
      "category",
    ],
  },
  "oldData": {
    "data": [
      {
        "data": [
          {
            "value": "1d6b2882-9a26-41d6-a976-dff75dfdd5e2",
          },
          {
            "value": "2004-05-11",
          },
          {
            "value": "3.93",
          },
          {
            "value": "PT",
          },
          {
            "value": "Category C",
          },
        ],
        "isEmphasized": false,
      },
      {
        "data": [
          {
            "value": "1d6b2f8c-13c7-4c33-9835-d7e9e5448f81",
          },
          {
            "value": "2012-12-16",
          },
          {
            "value": "2.18",
          },
          {
            "value": "AF",
          },
          {
            "value": "Category C",
          },
        ],
        "isEmphasized": false,
      },
      {
        "data": [
          {
            "value": "1d6b412b-e1be-4566-b27f-d6d2591be73e",
          },
          {
            "value": "2019-11-01",
          },
          {
            "value": "4.0",
          },
          {
            "value": "SI",
          },
          {
            "value": "Category A",
          },
        ],
        "isEmphasized": false,
      },
      {
        "data": [
          {
            "value": "1d6c2c23-39cb-418a-9e09-e22b8ca3601f",
          },
          {
            "value": "1993-10-25",
          },
          {
            "isEmphasized": true,
            "value": "2.59",
          },
          {
            "value": "PE",
          },
          {
            "value": "Category B",
          },
        ],
        "isEmphasized": true,
      },
      {
        "data": [
          {
            "value": "1d6c6e9c-3f2b-4f36-af8f-4ece80e1c456",
          },
          {
            "value": "2021-11-07",
          },
          {
            "value": "1.72",
          },
          {
            "value": "ME",
          },
          {
            "value": "Category C",
          },
        ],
        "isEmphasized": false,
      },
      {
        "data": [
          {
            "value": "1d6cd9a0-4f30-452f-8dda-34777e2814e0",
          },
          {
            "value": "2002-01-04",
          },
          {
            "value": "7.89",
          },
          {
            "value": "AM",
          },
          {
            "value": "Category C",
          },
        ],
        "isEmphasized": false,
      },
      {
        "data": [
          {
            "value": "1d6e65d0-0b9b-4b76-9517-ed65dd038b14",
          },
          {
            "value": "2009-01-25",
          },
          {
            "value": "4.63",
          },
          {
            "value": "GW",
          },
          {
            "value": "Category C",
          },
        ],
        "isEmphasized": false,
      },
      {
        "data": [],
        "isEllipsis": true,
      },
      {
        "data": [
          {
            "value": "3d1634ea-2648-41cf-8bc4-02063d57d6d4",
          },
          {
            "value": "2015-05-19",
          },
          {
            "value": "8.43",
          },
          {
            "value": "KE",
          },
          {
            "value": "Category A",
          },
        ],
        "isEmphasized": false,
      },
      {
        "data": [
          {
            "value": "3d16de2d-c626-46b7-90bb-211be6e38f31",
          },
          {
            "value": "2004-02-27",
          },
          {
            "value": "0.69",
          },
          {
            "value": "JO",
          },
          {
            "value": "Category A",
          },
        ],
        "isEmphasized": false,
      },
      {
        "data": [
          {
            "value": "3d16e1f2-a2fc-4b6a-a00b-fbc4e16aa485",
          },
          {
            "value": "2012-05-20",
          },
          {
            "value": "2.17",
          },
          {
            "value": "CZ",
          },
          {
            "value": "Category C",
          },
        ],
        "isEmphasized": false,
      },
    ],
    "diffType": "removed",
    "headers": [
      "unique_key",
      "date",
      "metric_value",
      "country_code",
      "category",
    ],
  },
}
`;

exports[`parsePatch should handle when headers change 1`] = `
{
  "newData": {
    "data": [
      {
        "data": [
          {
            "value": "2022-01-01__FR",
          },
          {
            "value": "2022-01-01",
          },
          {
            "value": "FR",
          },
          {
            "isEmphasized": true,
            "value": "22.22",
          },
          {
            "isEmphasized": true,
            "value": "-22.0",
          },
          {
            "isEmphasized": true,
            "value": "-22.0",
          },
          {
            "isEmphasized": true,
            "value": "-22.0",
          },
          {
            "isEmphasized": true,
            "value": "22.22",
          },
        ],
        "isEmphasized": true,
      },
    ],
    "diffType": "added",
    "headers": [
      "unique_key",
      "date_month",
      "country",
      "mrr_bop",
      "resurrected_mrr",
      "downsell_mrr",
      "fail_to_engage_mrr",
      "churn_mrr",
      "mrr_eop",
    ],
  },
  "oldData": {
    "data": [
      {
        "data": [
          {
            "value": "2022-01-01__FR",
          },
          {
            "value": "2022-01-01",
          },
          {
            "value": "FR",
          },
          {
            "isEmphasized": true,
            "value": "11.11",
          },
          {
            "isEmphasized": true,
            "value": "11.0",
          },
          {
            "isEmphasized": true,
            "value": "-11.0",
          },
          {
            "isEmphasized": true,
            "value": "-11.0",
          },
          {
            "isEmphasized": true,
            "value": "-11.0",
          },
          {
            "isEmphasized": true,
            "value": "11.11",
          },
        ],
        "isEmphasized": true,
      },
    ],
    "diffType": "removed",
    "headers": [
      "unique_key",
      "date_month",
      "country",
      "mrr_bop",
      "resurrected_mrr",
      "downsell_mrr",
      "fail_to_engage_mrr",
      "churn_mrr",
      "mrr_eop",
    ],
  },
}
`;

exports[`parsePatch should handle when headers reverse order 1`] = `
{
  "newData": {
    "data": [
      {
        "data": [
          {
            "value": "2022-01-01__FR",
          },
          {
            "value": "2022-01-01",
          },
          {
            "isEmphasized": true,
            "value": "22.22",
          },
          {
            "value": "FR",
          },
        ],
        "isEmphasized": true,
      },
    ],
    "diffType": "added",
    "headers": [
      "unique_key",
      "date_month",
      "mrr_bop",
      "country",
    ],
  },
  "oldData": {
    "data": [
      {
        "data": [
          {
            "value": "2022-01-01__FR",
          },
          {
            "value": "2022-01-01",
          },
          {
            "value": "FR",
          },
          {
            "isEmphasized": true,
            "value": "11.11",
          },
        ],
        "isEmphasized": true,
      },
    ],
    "diffType": "removed",
    "headers": [
      "unique_key",
      "date_month",
      "country",
      "mrr_bop",
    ],
  },
}
`;
